### DISCUSSION TITLE
Hifitime V4

### DISCUSSION THREAD

**Discussion by gwbres** [nyx-space/hifitime#258] [Ideas]
Hello @ChristopherRabotin,

it's fair to say my libs now make advanced usage of Hifitime (and now basic nyx-space usage also).
Especially since i have been working on an RTK solver, because it requires precise calculations in GNSS timescales.

I have to say a hypothetical hifitime V4 is not mandatory to my geodesics requirements since I can workaround my way with v3.8.4.

I'm opening this thread because it might be interesting for you to keep in mind how I work my way around : 

1. `Epoch::from_str(".... {}", ts)` is my goto method to create correct Epochs, whatever the timescale.
2. In RINEX we arrive with Gregorian datetimes expressed as Y, M, D, HH, MM, SS, NS in a specific timescale.
For a lazy implementation on our side, we would need something like `Epoch::from_gregorian_in_ts(y, m, d... ts)`.
For me such method is not impossible. It's basically  `Epoch::from_gregorian(y, m, d... )` but when you store the actual duration, you need to substract the delta with the given timescale initial datetime, and propagate the correct carry, as you already do in From_gregorian.
3. Unlocking IRNSST will enable RTK solving with only IRNSS vehicles in sight and/or expressing RTK results against IRNSST.
I'd say it's good if we have it in the future but not urgent.
4. v3.8.4 already allows RTK solving with Glonass vehicles although we don't support GLONASST.
That's because GLONASST is actually UTC+3 and actually offset back to UTC+0 in RINEX format. 
Therefore we already have means to express that. It will be cleaner to express what it actually is in the future, but again not urgent
5. I currently express all QZSS vehicles as GPST, since I think QZSST is locked to GPST. But i'm not 100% sure this is correct, I will need to read on that in the future. But I want to conclude Galileo, BeiDou and Glonass first

**Comment by ChristopherRabotin** [comment]
Hi Guillaume,

You're absolutely a power user of hifitime, and your feedback is super important. Thanks for the work around as well, I hope this will help other potential power users.

I just realized yesterday that NAIF defines J2000 as being at noon and not midnight: this would explain the issue I was seeing a few months ago on the V4 branch where there was exactly a twelve hour difference between what I expected to see and the actual result. Given this, I'll pick up the v4 work soon from your active branch and hopefully merge that into the 4.0-dev branch after fixing some of the tests.

I'm still working a lot on ANISE, and hifitime v4 is next on my to-do list.
